src/backend/distributed/executor/multi_utility.c

```diff
@@ -1045,88 +1056,203 @@ ExecuteDistributedDDLCommand(Oid relationId, const char *ddlCommandString)
+ * The resulting hash table maps shardIds to ShardConnection structs.
+ */
+static HTAB *
+OpenConnectionsToAllShardPlacements(List *shardIdList, char *relationOwner)
+{
+	HTAB *shardConnectionHash = CreateShardConnectionHash();
+	ListCell *shardCell = NULL;
+	MemoryContext oldContext = MemoryContextSwitchTo(TopTransactionContext);
+
+	foreach(shardCell, shardIdList)
+	{
+		uint64 *shardIdPointer = (uint64 *) lfirst(shardCell);
+		uint64 shardId = (*shardIdPointer);
+
+		OpenConnectionsToShardPlacements(shardId, shardConnectionHash, relationOwner);
 	}
 
-	return true;
+	shardPlacementConnectionList = ConnectionList(shardConnectionHash);
+
+	MemoryContextSwitchTo(oldContext);
+
+	if (!isXactCallbackRegistered)
+	{
+		RegisterXactCallback(CompleteShardPlacementTransactions, NULL);
+		isXactCallbackRegistered = true;
+	}
+
+	return shardConnectionHash;
+}
+
+
+/*
+ * CompleteShardPlacementTransactions commits or aborts pending shard placement
+ * transactions when the local transaction commits or aborts.
+ */
+static void
+CompleteShardPlacementTransactions(XactEvent event, void *arg)
+{
+	if (shardPlacementConnectionList == NIL)
+	{
+		/* nothing to do */
+		return;
+	}
+	else if (event == XACT_EVENT_PRE_COMMIT)
+	{
+		/*
+		 * Any failure here will cause local changes to be rolled back,
+		 * and remote changes to either roll back (1PC) or, in case of
+		 * connection or node failure, leave a prepared transaction
+		 * (2PC).
+		 */
+
+		PrepareRemoteTransactions(shardPlacementConnectionList);
+
+		return;
+	}
+	else if (event == XACT_EVENT_COMMIT)
+	{
+		/*
+		 * A failure here will cause some remote changes to either
+		 * roll back (1PC) or, in case of connection or node failure,
+		 * leave a prepared transaction (2PC). However, the local
+		 * changes have already been committed.
+		 */
+
+		CommitRemoteTransactions(shardPlacementConnectionList, false);
+	}
+	else if (event == XACT_EVENT_ABORT)
+	{
+		/*
+		 * A failure here will cause some remote changes to either
+		 * roll back (1PC) or, in case of connection or node failure,
+		 * leave a prepared transaction (2PC). The local changes have
+		 * already been rolled back.
+		 */
+
+		AbortRemoteTransactions(shardPlacementConnectionList);
+	}
+	else
+	{
+		return;
+	}
+
+	CloseConnections(shardPlacementConnectionList);
+	shardPlacementConnectionList = NIL;
+}
+
+
+static void
+ExecuteCommandOnShardReplicas(StringInfo applyCommand, uint64 shardId,
+							  ShardConnections *shardConnections)
+{
+	List *connectionList = shardConnections->connectionList;
+	ListCell *connectionCell = NULL;
+
+	Assert(connectionList != NIL);
+
+	foreach(connectionCell, connectionList)
+	{
+		TransactionConnection *transactionConnection =
+			(TransactionConnection *) lfirst(connectionCell);
+		PGconn *connection = transactionConnection->connection;
+		PGresult *result = NULL;
+
+		/* send the query */
+		result = PQexec(connection, "BEGIN");
```

Comment:
I think it might make more sense to send the BEGIN after opening the connection. It creates a bit more symmetry between open+begin and commit+close which will help when we add support for multi-statement transactions.